src/services/EdFi.DataManagementService.Api/Core/Validation/EqualityConstraintValidator.cs

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Text.Json.Nodes;
+using EdFi.DataManagementService.Api.Core.Model;
+
+namespace EdFi.DataManagementService.Api.Core.Validation;
+
+public interface IEqualityConstraintValidator
+{
+    /// <summary>
+    /// Validates the equality constraints defined in MetaEd model are correct for the given API body.
+    /// </summary>
+    /// <param name="documentBody"></param>
+    /// <param name="equalityConstraints"></param>
+    /// <returns>Returns a list of validation failure messages.</returns>
+    IEnumerable<string>? Validate(JsonNode? documentBody, IEnumerable<EqualityConstraint> equalityConstraints);
+}
+
+public class EqualityConstraintValidator : IEqualityConstraintValidator
+{
+    public IEnumerable<string> Validate(JsonNode? documentBody, IEnumerable<EqualityConstraint> equalityConstraints)
+    {
+        var errors = new List<string>();
+        foreach (var equalityConstraint in equalityConstraints)
+        {
+
+            var sourcePath = Json.Path.JsonPath.Parse(equalityConstraint.SourceJsonPath.Value);
+            var targetPath = Json.Path.JsonPath.Parse(equalityConstraint.TargetJsonPath.Value);
+
+            var sourcePathResult = sourcePath.Evaluate(documentBody);
+            var targetPathResult = targetPath.Evaluate(documentBody);
+
+            var sourceValues = sourcePathResult.Matches!.Select(s => s.Value);
+            var targetValues = targetPathResult.Matches!.Select(t => t.Value);
+
+            if (!AllEqual(sourceValues.Concat(targetValues).ToList()))
```

Comment:
I tried to copy the logic from meadowlark here. Couple of questions though. What if source or target values are null? Are equality constraints always expected to be on required values?

Reply:
As mentioned above, the path may or may not exist. If it doesn't, don't worry about it. Prior JSON Schema validation means it's optional (or an ApiSchema.json error, but there needs to be some trust)